File: 11M2W/ehowlqk/boj_1167.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "boj_1167.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

let idx = 0;
const v = Number(input[idx++]);

const edges = Array.from({ length: v + 1 }, () => []);

// 간선처리
for (idx; idx < input.length; idx++) {
  const tmp = input[idx].split(" ").map((e) => Number(e));

  let node = 1;
  while (tmp[node] > 0) {
    edges[tmp[0]].push([tmp[node], tmp[node + 1]]);
    node += 2;
  }
}

// start로부터 가장 먼 노드와 거리 찾기
function findFurthest(start, edges, v) {
  visited = Array(v + 1).fill(false);
  visited[start] = true;
  const stack = [[start, 0]];

  let maxDist = 0,
    furthest = start;

  while (stack.length) {
    const [node, cost] = stack.pop();

    if (cost > maxDist) {
      maxDist = cost;
      furthest = node;
    }

    for (const [nxt, dist] of edges[node]) {
      if (!visited[nxt]) {
        stack.push([nxt, cost + dist]);
        visited[nxt] = true;
      }
    }
  }

  return [maxDist, furthest];
}

const EoD = findFurthest(1, edges, v)[1];
console.log(findFurthest(EoD, edges, v)[0]);

File: 11M2W/ehowlqk/boj_1167.py
import sys
sys.setrecursionlimit(10**6)

input = open('boj_1167.txt', 'r').readline
# input = sys.stdin.readline

def find_furthest(start, edges, v):
    visited = [False] * (v+1)
    visited[start] = True
    def bfs(node, cost):
        result = (node, cost)
        # print("visited:", node)
        for nxt, dist in edges[node]:
            if not visited[nxt]:
                # print("edges:", nxt, dist)
                visited[nxt] = True
                tmp = bfs(nxt, cost + dist)
                if tmp[1] > result[1]:
                    result = tmp
                visited[nxt] = False
        
        return result
    
    return bfs(start, 0)


v = int(input())

edges = [list() for _ in range(v+1)]

for _ in range(v):
    lst = list(map(int, input().split()))

    node = lst[0]
    for i in range(len(lst) // 2 - 1):
        edges[node].append((lst[2*i + 1], lst[2*i + 2]))

EoD = find_furthest(1, edges, v)[0]
print(find_furthest(EoD, edges, v)[1])



    



"""
2n + 1
2n+1, 2n+2

"""